src/evict.c

```diff
@@ -586,33 +587,35 @@ int performEvictions(void) {
             server.maxmemory_policy == MAXMEMORY_VOLATILE_TTL)
         {
             struct evictionPoolEntry *pool = EvictionPoolLRU;
+            dbKeyType keyType = (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS ?
+                                 DB_MAIN : DB_EXPIRES);
 
             while (bestkey == NULL) {
-                unsigned long total_keys = 0, keys;
+                unsigned long total_keys = 0, current_db_keys;
 
                 /* We don't want to make local-db choices when expiring keys,
                  * so to start populate the eviction pool sampling keys from
                  * every DB. */
                 for (i = 0; i < server.dbnum; i++) {
                     db = server.db+i;
+                    if ((current_db_keys = dbSize(db, keyType)) == 0) continue;
+
+                    total_keys += current_db_keys;
+                    unsigned long sampled_keys = 0;
                     do {
-                        int slot = 0;
-                        if (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS) {
-                            slot = getFairRandomSlot(db, DB_MAIN);
-                            dict = db->dict[slot];
-                        } else {
-                            slot = getFairRandomSlot(db, DB_EXPIRES);
-                            dict = db->expires[slot];
-                        }
-                        if ((keys = dictSize(dict)) != 0) {
-                            evictionPoolPopulate(i, slot, dict, db, pool);
-                            total_keys += keys;
+                        int slot = getFairRandomSlot(db, keyType);
+                        dict = (keyType == DB_MAIN ? db->dict[slot] : db->expires[slot]);
+                        if (dictSize(dict) != 0) {
+                            sampled_keys += evictionPoolPopulate(i, slot, dict, db->dict[slot], pool);
                         }
-                    /* Since keys are distributed across smaller slot-specific dictionaries in cluster mode, we may need to
-                     * visit more than one dictionary in order to populate required number of samples into eviction pool. */
-                    } while (server.cluster_enabled && keys != 0 && server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS &&
-                        total_keys < (unsigned long) server.maxmemory_samples
-                    );
+                    /* If there are a sufficient number of keys in the current database
+                     * (e.g., 10 times the value of maxmemory_samples), it is necessary
+                     * to ensure that enough keys are sampled.
+                     *
+                     * To avoid the situation where the number of keys in a single slot
+                     * in cluster mode is too low to meet the sampling requirements. */
+                    } while (current_db_keys > (unsigned long) server.maxmemory_samples*10 &&
```

Comment:
>we exit the loop if either we got enough samples from the current dict, or the total number of keys in all the DBs sampled so far doesn't have enough keys in it.

I renamed variable `keys` to `current_db_keys`. And `sampled_keys` is also the sampling count for the entire current DB.

My idea is that if the current DB has a sufficient number of keys (10 times the `server.maxmemory_samples`), we need to ensure that at least `server.maxmemory_samples` are sampled in the current DB. Otherwise, we should exit as soon as possible to avoid excessive looping when the key distribution is sparse.

The new comments also explain the exit conditions.

Reply:
ok, i think i get it now.
i refactored the code without really changing it, let me know if it looks ok.

dropped the condition after calling evictionPoolPopulate since it's comment says `This function guarantees that it returns a slot whose dict is non-empty, unless the entire db is empty`.
`getFairRandomSlot >` vs `getFairRandomSlot >=` (when negating the condition) isn't significant in this case.

Reply:
looks good : )

Reply:
Good summary 👍